ApollosProject/apollos-prototype
# Adds GraphQL Schema Diff and Linting

## DESCRIPTION

<img width="815" alt="Screen Shot 2019-05-31 at 10 21 02 PM" src="https://user-images.githubusercontent.com/2659478/58742368-6d43e080-83f2-11e9-824d-4cdf26f0d99c.png">

<img width="563" alt="Screen Shot 2019-05-31 at 10 50 07 PM" src="https://user-images.githubusercontent.com/2659478/58742919-8189dc80-83f6-11e9-9663-4d94dc310f50.png">


### What does this PR do, or why is it needed?

This PR does two things:
1. Gives us the ability to examine when the schema is updated and test for breaking changes via `yarn` and Travis
2. Enables client-side query lint checking against the schema.

### What design trade-offs/decisions were made?

Added a couple scripts and config files. Also adds a new stage in Travis. Added four new packages:
- `graphql-schema-linter`
- `get-graphql-schema`
- `@entria/graphql-findbreakingchanges`
- `eslint-plugin-graphql`

The first three could all be replaced with Prisma's `graphql-cli` except you invoke it with `graphql` and we already have a package called `graphql` so it throws an error 😞 

### How do I test this PR?

1. Make a breaking change to the schema (ex. change a field name)
2. `yarn schema-diff`
3. Write some bad code to a query in the app and make sure `eslint` yells at you.

### What automated tests did you write?

None.

### Possible Next Steps

1. **Fix schema formatting**. I added a `yarn schema-lint` command as well, very insightful but right now there are 328 errors, mostly just missing descriptions but possibly a few good ones as well. Maybe down the road, we can fix these and include the command in `yarn lint` for CI.
2. **Enable tracing**. This PR makes it impossible to commit breaking changes so we'll have to mark fields deprecated to slowly encourage client changes. When we do get ready to remove fields, we'll need a way to measure impact. This can be done by enabling tracing and acting on the recorded response data. 

## TODO:

- [x] Closes [#744]
- [ ] fix current lint errors
- [x] update `schema.graphql` dynamically

## REVIEW:

**Manual QA**


**Code Review: Questions to consider**


> The purpose of PR Review is to _improve the quality of the software._
